### ISSUE TITLE
missing OrderedDict in taurus.external.ordereddict

### ISSUE THREAD

**Issue by jkotan** [taurus-org/taurus#592]
Hi, 

I've just tried to test taurus on stretch and because of missing workaround for a pyqt bug in Taurus 4.0 (original package for stretch is not usable, i.e. macrogui fails) I decide to go to Taurus 4.1.

Unfortunately, when I try to run sardana I get the following error:

    Traceback (most recent call last):
      File "/usr/bin/Pool", line 11, in <module>
	load_entry_point('sardana==2.2.2', 'console_scripts', 'Pool')()
      File "/usr/lib/python2.7/dist-packages/sardana/tango/pool/__init__.py", line 101, in main
	run(start_time=datetime.datetime.now())
      File "/usr/lib/python2.7/dist-packages/sardana/tango/pool/__init__.py", line 95, in main_pool
	name=SERVER_NAME)
      File "/usr/lib/python2.7/dist-packages/sardana/tango/core/util.py", line 965, in run
	prepare_func(tango_util)
      File "/usr/lib/python2.7/dist-packages/sardana/tango/pool/__init__.py", line 75, in prepare_pool
	from .Pool import PoolClass, Pool
      File "/usr/lib/python2.7/dist-packages/sardana/tango/pool/Pool.py", line 45, in <module>
	from sardana.pool.pool import Pool as POOL
      File "/usr/lib/python2.7/dist-packages/sardana/pool/pool.py", line 56, in <module>
	from sardana.pool.poolcontrollermanager import ControllerManager
      File "/usr/lib/python2.7/dist-packages/sardana/pool/poolcontrollermanager.py", line 40, in <module>
	from taurus.external.ordereddict import OrderedDict
    ImportError: No module named ordereddict
    terminate called after throwing an instance of 'omni_thread_fatal'

Is Taurus 4.0 to 4.1 backward compatible or `lib/taurus/external/ordereddict/__init__.py` is missing?

The following fix

    --- /dev/null
    +++ b/lib/taurus/external/ordereddict/__init__.py
    @@ -0,0 +1,8 @@
    +from __future__ import absolute_import
    +
    +try:
    +    # ordereddict from python 2.7 or from ordereddict installed package?
    +    from collections import OrderedDict
    +except ImportError:
    +    # ordereddict from local import
    +    import os

works for me.

Anyhow, it would be good to update stretch packages for taurus/sardana to working versions.

Bests,
Jan

**Comment by cmft** [comment]
Hi @jkotan ,
OrderDict class is provided by `collections`library, so  we decided to remove it from Taurus (4.1.0)[https://github.com/taurus-org/taurus/blob/develop/CHANGELOG.md#410---2017-07-21].
Sardana (2.3.0)[https://github.com/sardana-org/sardana/blob/develop/CHANGELOG.md#230---2017-08-11] has  been changed to not use this dependency from Taurus 

If i'm not wrong the sardana and taurus packages can not be backporting to `stretch` since there is not any `debian security` problem.

IMHO  the solution is update your sardana package

Best,
Carlos

**Comment by cpascual** [comment]
> Anyhow, it would be good to update stretch packages for taurus/sardana to working versions.

As @cmft  says, we cannot update the official packages in stretch. But we can update them for sid and provide backports, which is in our TODO-list

**Comment by cpascual** [comment]
Since this is not a bug in taurus, i am closing the issue as rejected.

**Comment by jkotan** [comment]
Thanks. Ok, I see. So you allows for backward incompatibility in changes of minor versions. Good to know.

**Comment by cpascual** [comment]
> So you allows for backward incompatibility in changes of minor versions.

@jkotan : You are right in pointing this out. In principle we should not, since we adhere to [Semantic Versioning](http://semver.org/).

When we released 4.1, we saw this as removing a third party bundled library that was already provided by the dependencies and thus it was not enough to qualify for a major version change... but your comment made me realize that we are not only removing the bundled code, but also a taurus API (and this causes backwards incompatibility).

Therefore I propose to reopen this as a bug and fix it by providing the API (with a deprecation warning) and **without** bundling the library (since we now depend on python 2.7 and therefore ordereddict should be available in the collections module).
